# What is random walk?
A random walk is the process by which randomly-moving objects wander away from where they started. Yes, that is it, easy-peasy.

Lets understand this with the help of an example.
Suppose you have a unbiased(having 50:50 chances of Heads and Tails) coin. You are going to flip it 100 times and if you get a Heads you will move forward a step and on the other hand if you get a Tails you will move a step back. And after completing 100 flips see how far backwards or forwards are you from the starting point.

Now you will thinking, this is mundane and probably this has no practical use. And I must confess I thought the exact same.

But to our surprise **Random Walk** has a plethora of applications and have gave rise to make mathematical and statistical studies.<br>
The path traced by a molecule as it travels in a liquid or a gas, the search path of a foraging animal, the price of a fluctuating stock and the financial status of a gambler: all can be approximated by random walk models, even though they may not be truly random in reality.

For more info please conside these resources. <a href='https://www.mit.edu/~kardar/teaching/projects/chemotaxis(AndreaSchmidt)/random.htm'>[1]</a>, [[2]](https://www.khanacademy.org/computing/computer-programming/programming-natural-simulations/programming-randomness/a/random-walks), [[3]](https://en.wikipedia.org/wiki/Random_walk)


## Random Walk implementation (python)

The main aim of writing this notebook is to understand the effifciency of Vectoriation. And Vectorization is technique of representing data in forms of vectors. And arithmetic operations on vectors help us compute data simultaneously rather than on repeating an operation. Confused? take a look the example below and revist the theory.

#### We can implement Random walk in two ways: 
* Classic looping way
* Vectorized method using (Numpy)

### Classic way (using good old loops)

In [1]:
import random


def random_walk_classic(n_steps = 10000):
    pos = 0                                         # origin point
    walk = [pos]                                    # list to keep track of walk
    for x in range(n_steps):
        step = 1 if random.randint(0,1) else -1     # generates psuedorandom coin flip outcomes
        pos += step
        walk.append(pos)
    return walk, pos

In [2]:
_, pos = random_walk_classic()                 # here _ is used as a throwaway variable
                                               # to catch the return value.
print("Final Point :", pos)

Final Point : -52


### Vectorised Way (using Numpy) 

In [3]:
import numpy as np


def random_walk_numpy(nsteps = 10000):
    draws = np.random.randint(0, 2, size=nsteps)     # Generates a 1D array of psuedorandom
                                                     # coin flip outcomes
    steps = np.where(draws > 0, 1, -1)               # Array of steps taken on each flip
                                                     # (either +1 or -1) 
    walk = steps.cumsum()                            # Array of cummulative sum of steps
    return walk

In [4]:
walk = random_walk_numpy()                 # here _ is used as a throwaway variable
                                           # to catch the return value.
print("Final Point :", walk[-1])

Final Point : -148


### Let's check the efficiency for each one 

In [5]:
%timeit _ = random_walk_classic()

7.63 ms ± 38 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)


In [6]:
%timeit _ = random_walk_numpy()

138 µs ± 993 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)


## Verdict
Now classic method took approximately 7.63 ms<br>
And numpy method approximately 138 µs

1 milli unit = 1000 micro unit, therefore vectorised method(numpy) is approximately 55 time faster in execution than classic loop method.

## Credits:
I got the idea and the example out of [Wes McKinney's](https://en.wikipedia.org/wiki/Wes_McKinney) book Python For Data Analysis. And highly recommend this [book](https://www.amazon.com/gp/product/B075X4LT6K/ref=dbs_a_def_rwt_bibl_vppi_i0) if you are into Data Analytics.

Also I have attached 3 links the top-most cell, please have quick eye of them to.